allow defining negative conditional factors with a leading !

jurko-gospodnetic opened the pull request:
Pull request created for discussion on tox-dev/tox issues #292 and #688.

Still missing a doc/changelog/contributors changes, but the implementation & tests should be ok.

Suggested discussion points:
- negated factor operator other than `!`?
- any missing tests?

jurko-gospodnetic commented:
This is a replacement for the original pull request https://github.com/jurko-gospodnetic/tox/pull/1 created in the fix repo before all the requirement pull requests got merged.

gaborbernat commented:
@jurko-gospodnetic the code style check failed

jurko-gospodnetic commented:
@gaborbernat - heh, yeah, as the pull request description said at the time: `and there might be fix-lint failures to clean up` ;)

Should be fixed now.

gaborbernat commented:
@asottile not sure what the pyupgrade means, but it failed.

asottile commented:
in this case it's probably rewriting the python<2.7 set call to a python2.7+ set comprehension

gaborbernat commented on tests/test_config.py:
```diff
@@ -1376,11 +1376,16 @@ def test_factors(self, newconfig):
         configs = conf.envconfigs
-        assert [dep.name for dep in configs['a-x'].deps] == ["dep-all", "dep-a", "dep-x"]
-        assert [dep.name for dep in configs['b'].deps] == ["dep-all", "dep-b"]
+        assert [dep.name for dep in configs['a-x'].deps] == \
```
I prefer avoiding using the line separator - instead consider setting out the right hand side to a local variable e.g. ``exp_deps``.

jurko-gospodnetic replied:
ok, changing it to:
```
        expected = ["dep-all", "dep-a", "dep-x", "dep-!b"]
        assert [dep.name for dep in configs['a-x'].deps] == expected
        expected = ["dep-all", "dep-b", "dep-!a", "dep-!x"]
        assert [dep.name for dep in configs['b'].deps] == expected
```
Though I'd personally prefer:
```
        assert [dep.name for dep in configs['a-x'].deps] == [
            "dep-all", "dep-a", "dep-x", "dep-!b"]
        assert [dep.name for dep in configs['b'].deps] == [
            "dep-all", "dep-b", "dep-!a", "dep-!x"]
```

gaborbernat commented on tox/config.py:
```diff
@@ -913,8 +913,20 @@ def _split_env(env):
+    def split_single(e):
+        raw = e.split('-')
+        included = {x for x in raw if x[:1] != '!'}
+        excluded = {x[1:] for x in raw if x[:1] == '!'}
```
this is a bit cryptic, consider instead using ``startswith``

gaborbernat commented on tox/config.py:
```diff
@@ -913,8 +913,20 @@ def _split_env(env):
 def _split_factor_expr(expr):
+    def split_single(e):
+        raw = e.split('-')
+        included = {x for x in raw if x[:1] != '!'}
```
this is a bit cryptic, consider instead using ``not startswith`` to make it more explicit

gaborbernat commented on tox/config.py:
```diff
@@ -913,8 +913,20 @@ def _split_env(env):
+def _split_factor_expr_all(expr):
     partial_envs = _expand_envstr(expr)
-    return [set(e.split('-')) for e in partial_envs]
+    return [{x[1:] if x[:1] == '!' else x for x in e.split('-')}
```
I feel like we should have a ``is_negated_factor`` function to avoid all this repetition of this check

gaborbernat replied:
the variable x is a bit too cryptic, e.g. the ``partial_envs`` hints that ``e`` will be a ``partial env``, but once I iterate on that what that contains?

jurko-gospodnetic commented:
@gaborbernat  - Ok, I updated the code with the changes you requested.

- take a look at `_is_negated_factor()` and `_base_factor_name()` functions and where they are used in `tox/config.py`
- added 2 introductory doc cleanup commits not directly related to the negated factor conditionals feature
- added negated factor conditional related changelog entry
- updated docs with negated factor conditional related info

I briefly played with implementing the included/excluded factor splitting using something like this:
```
        split = dict(itertools.groupby(raw, key=_is_negated_factor))
        included = set(split[False])
        excluded = set(split[True])
```
which makes the code more concise but potentially too cryptic. In the end kept it as two dict comprehensions. Feel free to tweak as you like it.

When merging this, I suggest keeping the initial doc cleanup commits separate, but I did not feel like opening a pull request just for those.

jurko-gospodnetic commented:
Btw. all this condition checking actually leads very nicely to a more expressive grammar supporting:
- `,` for `or`
- `-` for `and`
- `!` for `not`
- and `(`/`)` parenthesis for grouping

but that change is not precluded by just adding the ! operator support so I did not want to get into that in this pull request.

jurko-gospodnetic commented:
I'm not really sure what you want added to `config-v2.rst`. That document does not contain full specification of how factor conditions are checked, as exists in `config.rst`. It seems like there's a lot of technical debt there that would need to be paid before extra documentation for this feature can be added and I don't really think all that should be done as a part of this pull request.

gaborbernat commented:
Resolves #292.
